p2p/protocol/autonatv2/autonat.go

```diff
@@ -0,0 +1,138 @@
+package autonatv2
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/libp2p/go-libp2p/core/event"
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/p2p/protocol/autonatv2/pbv2"
+	ma "github.com/multiformats/go-multiaddr"
+	manet "github.com/multiformats/go-multiaddr/net"
+	"golang.org/x/exp/rand"
+)
+
+var log = logging.Logger("autonatv2")
+
+type AutoNAT struct {
+	host          host.Host
+	sub           event.Subscription
+	ctx           context.Context
+	cancel        context.CancelFunc
+	wg            sync.WaitGroup
+	srv           *Server
+	cli           *Client
+	allowAllAddrs bool // for testing
+}
+
+func New(h host.Host, dialer host.Host, opts ...AutoNATOption) (*AutoNAT, error) {
+	s := defaultSettings()
+	for _, o := range opts {
+		if err := o(s); err != nil {
+			return nil, err
+		}
+	}
+	sub, err := h.EventBus().Subscribe(new(event.EvtLocalReachabilityChanged))
```

Comment:
I copied this logic from AutoNATv1. Do we need to disable the server in case the node is private? 
We can provide this service over a relay connection too since private nodes can dial out and check reachability. Even with amplification attack prevention this only takes 100kb.

Reply:
added a comment explaining the transition period.